File: src/main/java/org/jcodec/containers/mp4/demuxer/MP4Demuxer.java
package org.jcodec.containers.mp4.demuxer;

import static org.jcodec.common.Fourcc.free;
import static org.jcodec.common.Fourcc.ftyp;
import static org.jcodec.common.Fourcc.mdat;
import static org.jcodec.common.Fourcc.moov;
import static org.jcodec.common.Fourcc.wide;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import org.jcodec.common.Codec;
import org.jcodec.common.Demuxer;
import org.jcodec.common.DemuxerTrack;
import org.jcodec.common.DemuxerTrackMeta;
import org.jcodec.common.SeekableDemuxerTrack;
import org.jcodec.common.TrackType;
import org.jcodec.common.UsedViaReflection;
import org.jcodec.common.io.NIOUtils;
import org.jcodec.common.io.SeekableByteChannel;
import org.jcodec.containers.mp4.MP4TrackType;
import org.jcodec.containers.mp4.MP4Util;
import org.jcodec.containers.mp4.MP4Util.Movie;
import org.jcodec.containers.mp4.boxes.AudioSampleEntry;
import org.jcodec.containers.mp4.boxes.Box;
import org.jcodec.containers.mp4.boxes.HandlerBox;
import org.jcodec.containers.mp4.boxes.MovieBox;
import org.jcodec.containers.mp4.boxes.NodeBox;
import org.jcodec.containers.mp4.boxes.SampleEntry;
import org.jcodec.containers.mp4.boxes.SampleSizesBox;
import org.jcodec.containers.mp4.boxes.TrakBox;
import org.jcodec.platform.Platform;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * Demuxer frontend for MP4
 * 
 * @author The JCodec project
 * 
 */
public class MP4Demuxer implements Demuxer {

    private List<SeekableDemuxerTrack> tracks;
    private TimecodeMP4DemuxerTrack timecodeTrack;
    MovieBox movie;
    protected SeekableByteChannel input;

    // modifies h264 to conform to annexb and aac to contain adts header
    public static MP4Demuxer createMP4Demuxer(SeekableByteChannel input) throws IOException {
        return new MP4Demuxer(input);
    }

    // does not modify packets
    public static MP4Demuxer createRawMP4Demuxer(SeekableByteChannel input) throws IOException {
        return new MP4Demuxer(input) {
            @Override
            protected SeekableDemuxerTrack newTrack(TrakBox trak) {
                return new MP4DemuxerTrack(movie, trak, this.input);
            }
        };
    }

    private SeekableDemuxerTrack fromTrakBox(TrakBox trak) {
        SampleSizesBox stsz = NodeBox.findFirstPath(trak, SampleSizesBox.class, Box.path("mdia.minf.stbl.stsz"));
        if (stsz == null)
            return null;

        SampleEntry[] sampleEntries = NodeBox.findAllPath(trak, SampleEntry.class,
                new String[] { "mdia", "minf", "stbl", "stsd", null });
        boolean isPCM = sampleEntries.length != 0 ? (sampleEntries[0] instanceof AudioSampleEntry)
                && isPCMCodec(Codec.codecByFourcc(sampleEntries[0].getFourcc())) : false;

        if (stsz.getDefaultSize() != 0 && isPCM)
            return new PCMMP4DemuxerTrack(movie, trak, input);
        return newTrack(trak);
    }

    private boolean isPCMCodec(Codec codec) {
        return codec != null && codec.isPcm();
    }

    protected SeekableDemuxerTrack newTrack(TrakBox trak) {
        return new CodecMP4DemuxerTrack(new MP4DemuxerTrack(movie, trak, input));
    }

    MP4Demuxer(SeekableByteChannel input) throws IOException {
        this.input = input;
        tracks = new LinkedList<SeekableDemuxerTrack>();
        findMovieBox(input);
    }

    private void findMovieBox(SeekableByteChannel input) throws IOException {
        Movie mv = MP4Util.parseFullMovieChannel(input);
        if (mv == null || mv.getMoov() == null)
            throw new IOException("Could not find movie meta information box");
        movie = mv.getMoov();

        processHeader(movie);
    }

    private void processHeader(NodeBox moov) throws IOException {
        TrakBox tt = null;
        TrakBox[] trakBoxs = NodeBox.findAll(moov, TrakBox.class, "trak");
        for (int i = 0; i < trakBoxs.length; i++) {
            TrakBox trak = trakBoxs[i];
            SampleEntry se = NodeBox.findFirstPath(trak, SampleEntry.class,
                    new String[] { "mdia", "minf", "stbl", "stsd", null });
            if (se != null && "tmcd".equals(se.getFourcc())) {
                tt = trak;
            } else {
                SeekableDemuxerTrack trakBox = fromTrakBox(trak);
                if (trakBox != null) {
                    tracks.add(trakBox);
                }
            }
        }
        if (tt != null) {
            DemuxerTrack video = getVideoTrack();
            if (video != null)
                timecodeTrack = new TimecodeMP4DemuxerTrack(movie, tt, input);
        }
    }

    public static MP4TrackType getTrackType(TrakBox trak) {
        HandlerBox handler = NodeBox.findFirstPath(trak, HandlerBox.class, Box.path("mdia.hdlr"));
        return MP4TrackType.fromHandler(handler.getComponentSubType());
    }

    public DemuxerTrack getVideoTrack() {
        for (SeekableDemuxerTrack demuxerTrack : tracks) {
            DemuxerTrackMeta meta = demuxerTrack.getMeta();
            if (meta.getType() == TrackType.VIDEO)
                return demuxerTrack;
        }
        return null;
    }

    public MovieBox getMovie() {
        return movie;
    }

    @Override
    public List<SeekableDemuxerTrack> getTracks() {
        return new ArrayList<SeekableDemuxerTrack>(tracks);
    }

    @Override
    public List<DemuxerTrack> getVideoTracks() {
        ArrayList<DemuxerTrack> result = new ArrayList<DemuxerTrack>();
        for (SeekableDemuxerTrack demuxerTrack : tracks) {
            DemuxerTrackMeta meta = demuxerTrack.getMeta();
            if (meta.getType() == TrackType.VIDEO)
                result.add(demuxerTrack);
        }
        return result;
    }

    @Override
    public List<DemuxerTrack> getAudioTracks() {
        ArrayList<DemuxerTrack> result = new ArrayList<DemuxerTrack>();
        for (SeekableDemuxerTrack demuxerTrack : tracks) {
            DemuxerTrackMeta meta = demuxerTrack.getMeta();
            if (meta.getType() == TrackType.AUDIO)
                result.add(demuxerTrack);
        }
        return result;
    }

    public TimecodeMP4DemuxerTrack getTimecodeTrack() {
        return timecodeTrack;
    }

    @UsedViaReflection
    public static int probe(final ByteBuffer b) {
        ByteBuffer fork = b.duplicate();
        int success = 0;
        int total = 0;
        while (fork.remaining() >= 8) {
            long len = Platform.unsignedInt(fork.getInt());
            int fcc = fork.getInt();
            int hdrLen = 8;
            if (len == 1) {
                len = fork.getLong();
                hdrLen = 16;
            } else if (len < 8)
                break;
            if (fcc == ftyp && len < 64 || fcc == moov && len < 100 * 1024 * 1024 || fcc == free || fcc == mdat
                    || fcc == wide)
                success++;
            total++;
            if (len >= Integer.MAX_VALUE)
                break;
            NIOUtils.skip(fork, (int) (len - hdrLen));
        }

        return total == 0 ? 0 : success * 100 / total;
    }

    @Override
    public void close() throws IOException {
        input.close();
    }
}

File: src/main/java/org/jcodec/containers/mp4/demuxer/MP4Demuxer.kt
package org.jcodec.containers.mp4.demuxer

import org.jcodec.common.*
import org.jcodec.common.io.NIOUtils
import org.jcodec.common.io.SeekableByteChannel
import org.jcodec.containers.mp4.MP4TrackType
import org.jcodec.containers.mp4.MP4Util
import org.jcodec.containers.mp4.boxes.*
import org.jcodec.containers.mp4.boxes.Box.Companion.path
import org.jcodec.containers.mp4.boxes.NodeBox.Companion.findAll
import org.jcodec.containers.mp4.boxes.NodeBox.Companion.findAllPath
import org.jcodec.containers.mp4.boxes.NodeBox.Companion.findFirstPath
import org.jcodec.platform.Platform
import java.io.IOException
import java.nio.ByteBuffer
import java.util.*

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * Demuxer frontend for MP4
 *
 * @author The JCodec project
 */
open class MP4Demuxer internal constructor(protected var input: SeekableByteChannel) : Demuxer {
    private val tracks: MutableList<SeekableDemuxerTrack>
    var timecodeTrack: TimecodeMP4DemuxerTrack? = null
        private set
    var movie: MovieBox? = null
    private fun fromTrakBox(trak: TrakBox?): SeekableDemuxerTrack {
        val stsz = findFirstPath(trak, path("mdia.minf.stbl.stsz")) as SampleSizesBox?
        val sampleEntries = findAllPath(trak, arrayOf("mdia", "minf", "stbl", "stsd", null)).map { asSampleEntry(it) }
        val isPCM = (sampleEntries[0] is AudioSampleEntry
                && isPCMCodec(Codec.codecByFourcc(sampleEntries[0]!!.fourcc)))
        return if (stsz!!.defaultSize != 0 && isPCM) PCMMP4DemuxerTrack(movie, trak, input) else newTrack(trak)
    }

    private fun isPCMCodec(codec: Codec?): Boolean {
        return codec != null && codec.isPcm
    }

    protected open fun newTrack(trak: TrakBox?): SeekableDemuxerTrack {
        return CodecMP4DemuxerTrack(MP4DemuxerTrack(movie!!, trak, input))
    }

    @Throws(IOException::class)
    private fun findMovieBox(input: SeekableByteChannel) {
        val mv = MP4Util.parseFullMovieChannel(input)
        if (mv == null || mv.moov == null) throw IOException("Could not find movie meta information box")
        movie = mv.moov
        processHeader(movie)
    }

    @Throws(IOException::class)
    private fun processHeader(moov: NodeBox?) {
        var tt: TrakBox? = null
        val trakBoxs = findAll(moov, "trak").map { it as? TrakBox }
        for (i in trakBoxs.indices) {
            val trak = trakBoxs[i]
            val findFirstPath = findFirstPath(trak, arrayOf("mdia", "minf", "stbl", "stsd", null))
            val se = asSampleEntry(findFirstPath)
            if (se != null && "tmcd" == se.fourcc) {
                tt = trak
            } else {
                tracks.add(fromTrakBox(trak))
            }
        }
        if (tt != null) {
            val video = videoTrack
            if (video != null) timecodeTrack = TimecodeMP4DemuxerTrack(movie, tt, input)
        }
    }


    val videoTrack: DemuxerTrack?
        get() {
            for (demuxerTrack in tracks) {
                val meta = demuxerTrack.meta
                if (meta.type == TrackType.VIDEO) return demuxerTrack
            }
            return null
        }

    override fun getTracks(): List<SeekableDemuxerTrack> {
        return ArrayList(tracks)
    }

    override fun getVideoTracks(): List<DemuxerTrack> {
        val result = ArrayList<DemuxerTrack>()
        for (demuxerTrack in tracks) {
            val meta = demuxerTrack.meta
            if (meta.type == TrackType.VIDEO) result.add(demuxerTrack)
        }
        return result
    }

    override fun getAudioTracks(): List<DemuxerTrack> {
        val result = ArrayList<DemuxerTrack>()
        for (demuxerTrack in tracks) {
            val meta = demuxerTrack.meta
            if (meta.type == TrackType.AUDIO) result.add(demuxerTrack)
        }
        return result
    }

    @Throws(IOException::class)
    override fun close() {
        input.close()
    }

    companion object {
        val PROBE = object : DemuxerProbe {
            override fun probe(b: ByteBuffer): Int {
                return MP4Demuxer.Companion.probe(b)
            }
        }

        fun asSampleEntry(box: Box?) =
                box as? SampleEntry? ?: box?.reinterpret(SampleEntry(box.header)) as? SampleEntry?

        // modifies h264 to conform to annexb and aac to contain adts header
        @JvmStatic
        @Throws(IOException::class)
        fun createMP4Demuxer(input: SeekableByteChannel): MP4Demuxer {
            return MP4Demuxer(input)
        }

        // does not modify packets
        @JvmStatic
        @Throws(IOException::class)
        fun createRawMP4Demuxer(input: SeekableByteChannel): MP4Demuxer {
            return object : MP4Demuxer(input) {
                override fun newTrack(trak: TrakBox?): SeekableDemuxerTrack {
                    return MP4DemuxerTrack(movie!!, trak, this.input)
                }
            }
        }

        fun getTrackType(trak: TrakBox?): MP4TrackType {
            val handler = findFirstPath(trak, path("mdia.hdlr")) as HandlerBox?
            return MP4TrackType.fromHandler(handler!!.componentSubType)
        }

        @UsedViaReflection
        fun probe(b: ByteBuffer): Int {
            val fork = b.duplicate()
            var success = 0
            var total = 0
            while (fork.remaining() >= 8) {
                var len = Platform.unsignedInt(fork.int)
                val fcc = fork.int
                var hdrLen = 8
                if (len == 1L) {
                    len = fork.long
                    hdrLen = 16
                } else if (len < 8) break
                if (fcc == Fourcc.ftyp && len < 64 || fcc == Fourcc.moov && len < 100 * 1024 * 1024 || fcc == Fourcc.free || fcc == Fourcc.mdat || fcc == Fourcc.wide) success++
                total++
                if (len >= Int.MAX_VALUE) break
                NIOUtils.skip(fork, (len - hdrLen).toInt())
            }
            return if (total == 0) 0 else success * 100 / total
        }
    }

    init {
        tracks = LinkedList()
        findMovieBox(input)
    }

}